design/20240918.delayed-certificate-activation.md

```diff
@@ -0,0 +1,141 @@
+    - [Backward Compatibilty](#backward-compatibilty)
+    - [Test Plan](#test-plan)
+    - [Upgrading](#upgrading)
+    - [Misc](#misc)
+  - [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+This proposal introduces a mechanism to specify a future `NotBefore` time in cert-manager certificates, enabling delayed activation of issued certificates.
+
+## Motivation
+
+Currently, cert-manager creates the Certificate Signing Request (CSR) with `NotBefore` set to the current time (`time.Now()`).  This prevents use cases where a certificate needs to be issued in advance but only become valid at a specific future time. For example,
+
+- **Pre-provisioning Certificates:** Generating certificates for resources that will be deployed in the future, such as for a scheduled application rollout.
+- **Compliance Requirements:** Adhering to specific security policies that mandate a future validation start date for certificates.
+
+Supporting a future `NotBefore` time aligns with the capabilities of x509.CertificateRequest and provides greater flexibility in certificate management.
+
+### Goals
+
+- Extend the `Certificate` and `CertificateRequest` APIs to allow specifying a future `NotBefore` time.
+- Update the `CertificateRequest` controller to honor the specified NotBefore time when creating x509.CertificateRequests.
+- Maintain backward compatibility with existing certificate issuance workflows.
+
+### Non-Goals
+
+- This proposal does not aim to change the behavior of existing certificates without a `NotBefore` field.
+
+## Proposal
+
+Introduce a new optional field, `NotBefore`, to the spec of both `Certificate` and `CertificateRequest` resources. This field will allow users to explicitly define the start time of the certificate's validity period.
+
+### API changes
+
+```
+type CertificateSpec struct {
+  // EXISTING FIELDS
+  // ...
+
+  // NotBefore defines the time when the certificate becomes valid.
+  // If omitted or set to a time in the past, the certificate will be 
+  // valid immediately upon issuance.
+  // +optional
+  NotBefore *metav1.Time `json:"notBefore,omitempty"`
+}
+
+type CertificateRequestSpec struct {
+  // EXISTING FIELDS
+  // ...
+
+  // NotBefore defines the time when the certificate becomes valid.
+  // If omitted or set to a time in the past, the certificate will be 
+  // valid immediately upon issuance.
+  // Date must be specified in RFC3339 format.
+  // +optional
+  NotBefore *metav1.Time `json:"notBefore,omitempty"` 
+}
+```
+
+### Examples
+
+#### Issuance with Future NotBefore
+
+```
+apiVersion: cert-manager.io/v1
+kind: Certificate
+metadata:
+  name: example-cert
+spec:
+  secretName: example-cert-tls
+  issuerRef:
+    name: letsencrypt-prod
+    kind: ClusterIssuer
+  commonName: example.com
+  notBefore: "2024-10-01T10:00:00Z" 
+  duration: 2160h # 90 days
+```
+
+This example creates a certificate that will only become valid on October 1st, 2024, even if it is issued earlier.
```

Comment:
This example is in the past. It might make sense to push the example date a couple of decades into the future, or to change the end of the prose `even if it is issued earlier` to `even if it was requested on January 1st, 2023`.
```suggestion
This example creates a certificate that would only become valid on October 1st, 2024, even if it was requested on January 1st, 2023.
```